File: src/ch/d/peters/swisstraintimetable/CsvHandler.java
package ch.d.peters.swisstraintimetable;

import ch.d.peters.swisstraintimetable.interfaces.IAbstractFactory;
import ch.d.peters.swisstraintimetable.interfaces.IDataHandler;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * @author Daniel Peters
 * @version 1.2
 */
public class CsvHandler<T> implements IDataHandler<T> {
  private static final String EXTENSION = ".csv";
  private String file;
  private IAbstractFactory<T> factory;

  public CsvHandler(String fileName, IAbstractFactory<T> abstractFactory) {
    file = fileName;
    factory = abstractFactory;
  }

  @Override
  public List<T> get() {
    var list = new ArrayList<T>();
    try (var reader = new BufferedReader(new FileReader(file + EXTENSION))) {
      reader.lines().forEach(line -> list.add(factory.getInstance(line)));
    } catch (IOException e) {
      System.out.println("IO error");
    }
    return list;
  }

  @Override
  public void put(List<T> element) {

  }

  @Override
  public void put(T element) {

  }
}

File: src/ch/d/peters/swisstraintimetable/storage/CsvHandler.kt
package ch.d.peters.swisstraintimetable.storage

import ch.d.peters.swisstraintimetable.interfaces.IAbstractFactory
import ch.d.peters.swisstraintimetable.interfaces.IDataHandler

import java.io.BufferedReader
import java.io.FileReader

/**
 * @author Daniel Peters
 * @version 1.2
 */
class CsvHandler<T>(
    private val file: String,
    private val factory: IAbstractFactory<T>
) : IDataHandler<T> {
  companion object {
    const val EXTENSION = ".csv"
  }


  override fun get(): List<T> {
    val list = ArrayList<T>()
    BufferedReader(FileReader(file + EXTENSION)).use {
      it.lines().forEach { line -> list.add(factory.getInstance(line)) }
    }
    return list
  }

  override fun put(elements: List<T>) {

  }

  override fun put(element: T) {

  }
}
